packages/plugin-ext/src/main/browser/editor/untitled-resource.ts

```diff
@@ -14,39 +14,103 @@
+            if (!untitledResource) {
+                return createUntitledResource('', '', uri, this.fileResourceResolver);
+            } else {
+                return untitledResource;
+            }
         }
-        throw new Error(`scheme ${uri.scheme} is not '${Schemes.UNTITLED}'`);
     }
 }
 
 export class UntitledResource implements Resource {
+    private fileResource: FileResource;
 
-    constructor(public uri: URI, private content?: string) {
+    constructor(public uri: URI, private content?: string, fileResourceResolver?: FileResourceResolver) {
+        const that = this;
         resources.set(this.uri.toString(), this);
+        if (fileResourceResolver) {
+            fileResourceResolver.resolveWithoutSchemaCheck(uri).then(resolvedFileResource => {
+                that.fileResource = resolvedFileResource;
+            });
+        }
     }
 
-    readContents(options?: { encoding?: string | undefined; } | undefined): Promise<string> {
-        return Promise.resolve(this.content ? this.content : '');
+    async init(): Promise<void> {
+        if (this.fileResource) {
+            return this.fileResource.init();
+        }
     }
 
     dispose(): void {
         resources.delete(this.uri.toString());
     }
+
+    async readContents(options?: { encoding?: string | undefined; } | undefined): Promise<string> {
+        if (this.content) {
+            return Promise.resolve(this.content);
+        } else if (this.fileResource) {
+            return this.fileResource.readContents(options);
+        } else {
+            return Promise.resolve('');
+        }
+    }
+
+    async saveContents(content: string, options?: { encoding?: string, overwriteEncoding?: string }): Promise<void> {
+        if (this.fileResource) {
+            this.fileResource.saveContents(content, options);
+        }
+    }
+
+    async saveContentChanges(changes: TextDocumentContentChangeEvent[], options?: { encoding?: string, overwriteEncoding?: string }): Promise<void> {
+        if (this.fileResource) {
+            this.fileResource.saveContentChanges(changes, options);
+        }
+    }
+}
+
+export async function createUntitledResource(content?: string, language?: string, uri?: URI, fileResourceResolver?: FileResourceResolver): Promise<UntitledResource> {
+    let extension;
+    if (language) {
+        for (const lang of monaco.languages.getLanguages()) {
+            if (lang.id === language) {
+                if (lang.extensions) {
+                    extension = lang.extensions[0];
+                    break;
+                }
+            }
+        }
+    }
+    return new UntitledResource(uri ? uri : new URI().withScheme(Schemes.UNTITLED).withPath(`/Untitled-${index++}${extension ? extension : ''}`),
+        content, fileResourceResolver);
 }
 
-export function createUntitledResource(content?: string, language?: string): UntitledResource {
+export async function createUntitledURI(language?: string): Promise<URI> {
```

Comment:
`async`? it does not use any promises

Reply:
please make it sync and then revert clients to be sync as well

Reply:
done